src/Crypto.ts

```diff
@@ -1,150 +1,152 @@
+  // Sync members
+  private cipher: Cipher
+  private decipher: Decipher
+  // Crypto lib
+  private cryptoLib: CryptoInterface
   constructor(
-    key: Buffer | string,
-    initVector: Buffer = Buffer.from(crypto.randomBytes(16)),
-    algorithm: string = 'aes-256-gcm',
-    useWebWorkers: boolean = false
+    masterKey: Buffer,
+    cryptoLib: CryptoInterface,
+    useWebWorkers: boolean = false,
+    algorithm: AlgorithmGCM = 'aes-256-gcm'
   ) {
+    // TODO: check the strength of the master key!
+    this.masterKey = masterKey
     this.algorithm = algorithm
-    this.initVector = initVector
-    // TODO: generate salt ?
-    this.key = this.pbkdfSync(key)
-    this.cipher = crypto.createCipheriv(algorithm, this.key, this.initVector)
-    this.decipher = crypto.createDecipheriv(algorithm, this.key, this.initVector)
+    this.cryptoLib = cryptoLib
     // Async via Process or Web workers
     this.useWebWorkers = useWebWorkers
     if (this.useWebWorkers) {
       spawn(new Worker("./CryptoWorker")).then((worker) => {
-        this.cryptorWorker = worker
-        this.cryptorWorker.init(this.algorithm, this.key, this.initVector)
+        this.cryptoWorker = worker
       })
     }
   }
 
-
-  encryptSync(plainBuf: string | Buffer, initVector?: Buffer): Buffer {
-    if (initVector && (initVector !== this.initVector)) {
-      this.resetCipherSync(initVector!)
+  ///////////
+  // Reset //
+  ///////////
+  resetSync(masterKey: Buffer, initVector: Buffer, salt: Buffer, authTag?: Buffer) {
+    // Generate key
+    const key = this.cryptoLib.pbkdf2Sync(masterKey, salt, cryptoConstants.PBKDF_NUM_ITERATIONS, cryptoConstants.KEY_LEN, 'sha512')
+    this.cipher = this.cryptoLib.createCipheriv(this.algorithm, key, initVector)
+    this.decipher = this.cryptoLib.createDecipheriv(this.algorithm, key, initVector)
+    if (authTag) {
+      this.decipher.setAuthTag(authTag)
     }
-    return Buffer.from(this.cipher.update(plainBuf))
   }
 
-  async encrypt(plainBuf: string | Buffer, initVector: Buffer | undefined = undefined): Promise<Buffer> {
-    if (initVector && (initVector !== this.initVector)) {
-      this.resetCipher(initVector!)
-    }
+  ////////////
+  // Cipher //
+  ////////////
+  encryptBlockSync(blockBuffer: Buffer): Buffer {
+    // Random initialization vector
+    const initVector = this.cryptoLib.randomBytes(cryptoConstants.INIT_VECTOR_LEN)
 
-    let buffer: Buffer
-    if (this.useWebWorkers && this.cryptorWorker) {
-      buffer = await this.cryptorWorker.encryptBuf(plainBuf)
-    } else {
-      buffer = Buffer.from(this.cipher.update(plainBuf))
-    }
-    return buffer
-  }
+    // Random salt
+    const salt = this.cryptoLib.randomBytes(cryptoConstants.SALT_LEN)
 
-  decryptSync(cipherBuf: Buffer, initVector?: Buffer): Buffer {
-    if (initVector && (initVector !== this.initVector)) {
-      this.resetDecipherSync(initVector!)
-    }
+    // Reset cipher
+    this.resetSync(this.masterKey, initVector, salt)
```

Comment:
Main thing here is to add a comment explaining why it's important to "reinitialize" with a new key and `initVector`. It's important to cover all the edgecases of random access read/write and document the behaviour.

Reply:
Since the cipher and decipher have been moved to the encrypt and decrypt methods, this is no longer ambiguous. I've added documentation to the functions and a note in to make it a bit more obvious.